src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -1948,6 +2147,114 @@ open_common(int (*real_open)(const char *pathname, int oflags, ...), const char
+			goto out_compatible;
+
+		/* query file object through ioctl() */
+		rc = fetch_dfs_obj_handle(fd_kernel, dfs_mt, &dfs_obj);
+		if (rc != 0) {
+			DS_WARN(rc, "fetch_dfs_obj_handle() failed");
+			goto out_compatible;
+		}
+
+		/* Need to create a fake fd and associate with fd_kernel */
+		atomic_fetch_add_relaxed(&num_open, 1);
+		dfs_get_mode(dfs_obj, &mode_query);
+
+		/* regular file */
+		if (S_ISREG(mode_query)) {
+			rc = find_next_available_fd(NULL, &idx_fd);
+			if (rc)
+				goto out_compatible_release;
+			file_list[idx_fd]->dfs_mt    = dfs_mt;
+			file_list[idx_fd]->file      = dfs_obj;
+			file_list[idx_fd]->parent    = parent;
+			file_list[idx_fd]->st_ino    = FAKE_ST_INO(full_path);
+			file_list[idx_fd]->idx_mmap  = -1;
+			file_list[idx_fd]->open_flag = oflags;
+			file_list[idx_fd]->offset    = 0;
+			if (strncmp(full_path, "/", 2) == 0)
+				D_STRNDUP(file_list[idx_fd]->path, dfs_mt->fs_root, DFS_MAX_PATH);
+			else
+				D_ASPRINTF(file_list[idx_fd]->path, "%s%s", dfs_mt->fs_root, full_path);
+			if (file_list[idx_fd]->path == NULL) {
+				free_fd(idx_fd, false);
+				goto out_compatible;
+			}
+			strncpy(file_list[idx_fd]->item_name, item_name, DFS_MAX_NAME);
+			fd_fake = idx_fd + FD_FILE_BASE;
+		}
+
+		/* directory */
+		if (S_ISDIR(mode_query)) {
+			rc = find_next_available_dirfd(NULL, &idx_dirfd);
+			if (rc)
+				goto out_compatible_release;
+			dir_list[idx_dirfd]->dfs_mt   = dfs_mt;
+			dir_list[idx_dirfd]->fd       = idx_dirfd + FD_DIR_BASE;
+			dir_list[idx_dirfd]->offset   = 0;
+			dir_list[idx_dirfd]->dir      = dfs_obj;
+			dir_list[idx_dirfd]->num_ents = 0;
+			dir_list[idx_dirfd]->st_ino   = FAKE_ST_INO(full_path);
+			memset(&dir_list[idx_dirfd]->anchor, 0, sizeof(daos_anchor_t));
+			dir_list[idx_dirfd]->path = NULL;
+			dir_list[idx_dirfd]->ents = NULL;
+			if (strncmp(full_path, "/", 2) == 0)
+				full_path[0] = 0;
+
+			/* allocate memory for ents[]. */
+			D_ALLOC_ARRAY(dir_list[idx_dirfd]->ents, READ_DIR_BATCH_SIZE);
+			if (dir_list[idx_dirfd]->ents == NULL) {
+				free_dirfd(idx_dirfd);
+				goto out_compatible;
+			}
+			D_ASPRINTF(dir_list[idx_dirfd]->path, "%s%s", dfs_mt->fs_root, full_path);
+			if (dir_list[idx_dirfd]->path == NULL) {
+				free_dirfd(idx_dirfd);
+				goto out_compatible;
+			}
+			if (strnlen(dir_list[idx_dirfd]->path, DFS_MAX_PATH) >= DFS_MAX_PATH) {
+				DS_WARN(ENAMETOOLONG, "path is longer than DFS_MAX_PATH");
+				free_dirfd(idx_dirfd);
+				goto out_compatible;
+			}
+			fd_fake = idx_dirfd + FD_DIR_BASE;
+		}
+		/* assert(fd_fake >= FD_FILE_BASE); */
+
+		/* add fd_kernel to hash table */
+		D_ALLOC(fd_ht_obj, sizeof(*fd_ht_obj));
+		if (fd_ht_obj == NULL) {
+			if (fd_fake >= FD_DIR_BASE)
+				free_dirfd(idx_dirfd);
+			else
+				free_fd(idx_fd, false);
+			goto out_compatible;
+		}
+		fd_ht_obj->real_fd = fd_kernel;
+		fd_ht_obj->fake_fd = fd_fake;
+		rc = d_hash_rec_insert(fd_hash, &fd_ht_obj->real_fd, sizeof(int), &fd_ht_obj->entry,
+				       true);
```

Comment:
Setting `exclusive` to true here will mean an exhaustive hash table search for the entry, presumably that's not needed as the `real_fd` should already be unique.

`d_hash_req_insert` never returns non-zero if exclusive is false as well, so if you were happy passing in false then you can just assert on the result.

Reply:
You are right. "real_fd" is unique already. I will remove "true" for exclusive flag and add "D_ASSERT(rc == 0);" after d_hash_rec_insert(). Thank you!

Reply:
Thank you! It is updated as you suggested.